File: companion/lib/codexRollout.js
function codexToolArg(name, argsJson) {
  try {
    const a = JSON.parse(argsJson || '{}');
    if (a.command) return String(Array.isArray(a.command) ? a.command.join(' ') : a.command).split('\n')[0].slice(0, 40);
    if (a.path) return String(a.path).split('/').slice(-2).join('/');
    if (a.pattern) return String(a.pattern);
  } catch (_) {}
  return '';
}

function parseCodexRollout(text) {
  const r = {
    sessionId: '', cwd: '', cliVersion: '', originator: '', model: '', effort: '',
    contextWindow: 0, gitBranch: '', firstPrompt: '', currentTask: '',
    totalInput: 0, totalOutput: 0, totalCacheRead: 0, lastContextTokens: 0, turnCount: 0,
    fiveHourPct: null, fiveHourResetsAt: null, weeklyPct: null, weeklyResetsAt: null, done: false,
  };
  for (const line of String(text || '').split('\n')) {
    const s = line.trim(); if (!s) continue;
    let e; try { e = JSON.parse(s); } catch (_) { continue; }
    const p = e.payload || {};
    if (e.type === 'session_meta') {
      r.sessionId = p.id || r.sessionId; r.cwd = p.cwd || r.cwd;
      r.cliVersion = p.cli_version || r.cliVersion; r.originator = p.originator || r.originator;
      if (p.git && p.git.branch) r.gitBranch = p.git.branch;
    } else if (e.type === 'turn_context') {
      if (p.model) r.model = p.model;
      if (p.effort) r.effort = p.effort;
      if (p.model_context_window) r.contextWindow = Number(p.model_context_window) || r.contextWindow;
    } else if (e.type === 'response_item' && p.type === 'function_call') {
      r.currentTask = (String(p.name || '') + ' ' + codexToolArg(p.name, p.arguments)).trim();
    } else if (e.type === 'event_msg') {
      if (p.type === 'user_message') { r.turnCount += 1; if (!r.firstPrompt) r.firstPrompt = String(p.message || '').trim(); }
      else if (p.type === 'task_complete') r.done = true;
      else if (p.type === 'token_count') {
        const tot = (p.info && p.info.total_token_usage) || {};
        const last = (p.info && p.info.last_token_usage) || {};
        const cached = Number(tot.cached_input_tokens) || 0;
        r.totalInput = Math.max(0, (Number(tot.input_tokens) || 0) - cached);
        r.totalCacheRead = cached;
        r.totalOutput = Number(tot.output_tokens) || 0;
        if (last.input_tokens != null) r.lastContextTokens = Number(last.input_tokens) || 0;
        const rl = p.rate_limits || {};
        for (const slot of ['primary', 'secondary']) {
          const w = rl[slot]; if (!w) continue;
          const resets = w.resets_at ? Math.floor(Date.parse(w.resets_at) / 1000) || null : null;
          if (Number(w.window_minutes) <= 300) { r.fiveHourPct = Number(w.used_percent); r.fiveHourResetsAt = resets; }
          else { r.weeklyPct = Number(w.used_percent); r.weeklyResetsAt = resets; }
        }
      }
    }
  }
  if (!r.contextWindow) r.contextWindow = 272000;     // Codex 默认窗口兜底
  return r;
}

module.exports = { parseCodexRollout, codexToolArg };

File: companion/codey/codex_rollout.py
"""解析 Codex CLI rollout JSONL -> 聚合对象(含 rate_limits)。纯函数。

注意:Codex 的 input_tokens 已含 cached_input_tokens,故 total_input 存非缓存部分,
cache_read 单列;rate_limits 按 window_minutes<=300 分 5h / 周。
"""
import json

from .util import parse_ts_ms


def codex_tool_arg(name, args_json):
    try:
        a = json.loads(args_json or "{}")
    except (ValueError, TypeError):
        return ""
    if not isinstance(a, dict):
        return ""
    if a.get("command"):
        cmd = a["command"]
        cmd = " ".join(str(x) for x in cmd) if isinstance(cmd, list) else str(cmd)
        return cmd.split("\n")[0][:40]
    if a.get("path"):
        return "/".join(str(a["path"]).split("/")[-2:])
    if a.get("pattern"):
        return str(a["pattern"])
    return ""


def _int(v):
    try:
        return int(v or 0)
    except (TypeError, ValueError):
        return 0


def parse_codex_rollout(text):
    r = {
        "session_id": "", "cwd": "", "cli_version": "", "originator": "", "model": "", "effort": "",
        "context_window": 0, "git_branch": "", "first_prompt": "", "current_task": "",
        "total_input": 0, "total_output": 0, "total_cache_read": 0, "last_context_tokens": 0,
        "turn_count": 0, "five_hour_pct": None, "five_hour_resets_at": None,
        "weekly_pct": None, "weekly_resets_at": None, "done": False,
    }
    for line in str(text if text is not None else "").split("\n"):
        s = line.strip()
        if not s:
            continue
        try:
            e = json.loads(s)
        except (ValueError, TypeError):
            continue
        if not isinstance(e, dict):
            continue
        p = e.get("payload") or {}
        et = e.get("type")
        if et == "session_meta":
            r["session_id"] = p.get("id") or r["session_id"]
            r["cwd"] = p.get("cwd") or r["cwd"]
            r["cli_version"] = p.get("cli_version") or r["cli_version"]
            r["originator"] = p.get("originator") or r["originator"]
            git = p.get("git") or {}
            if git.get("branch"):
                r["git_branch"] = git["branch"]
        elif et == "turn_context":
            if p.get("model"):
                r["model"] = p["model"]
            if p.get("effort"):
                r["effort"] = p["effort"]
            if p.get("model_context_window"):
                r["context_window"] = _int(p["model_context_window"]) or r["context_window"]
        elif et == "response_item" and p.get("type") == "function_call":
            r["current_task"] = (str(p.get("name") or "") + " "
                                 + codex_tool_arg(p.get("name"), p.get("arguments"))).strip()
        elif et == "event_msg":
            pt = p.get("type")
            if pt == "user_message":
                r["turn_count"] += 1
                if not r["first_prompt"]:
                    r["first_prompt"] = str(p.get("message") or "").strip()
            elif pt == "task_complete":
                r["done"] = True
            elif pt == "token_count":
                info = p.get("info") or {}
                tot = info.get("total_token_usage") or {}
                last = info.get("last_token_usage") or {}
                cached = _int(tot.get("cached_input_tokens"))
                r["total_input"] = max(0, _int(tot.get("input_tokens")) - cached)
                r["total_cache_read"] = cached
                r["total_output"] = _int(tot.get("output_tokens"))
                if last.get("input_tokens") is not None:
                    r["last_context_tokens"] = _int(last.get("input_tokens"))
                rl = p.get("rate_limits") or {}
                for slot in ("primary", "secondary"):
                    w = rl.get(slot)
                    if not w:
                        continue
                    ms = parse_ts_ms(w.get("resets_at")) if w.get("resets_at") else 0
                    resets = ms // 1000 if ms else None
                    try:
                        win = float(w.get("window_minutes") or 0)
                    except (TypeError, ValueError):
                        win = 0
                    pct = w.get("used_percent")
                    pct = float(pct) if pct is not None else None
                    if win <= 300:
                        r["five_hour_pct"], r["five_hour_resets_at"] = pct, resets
                    else:
                        r["weekly_pct"], r["weekly_resets_at"] = pct, resets
    if not r["context_window"]:
        r["context_window"] = 272000
    return r
